src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs

```diff
@@ -183,22 +184,28 @@ public static partial class JsonSerializer
                 ThrowHelper.ThrowJsonException_DeserializeDataRemaining(totalBytesRead, bytesInBuffer);
             }
 
-            return (TValue)state.Current.ReturnValue;
+            return (TValue)readStack.Current.ReturnValue;
         }
 
         private static void ReadCore(
             ref JsonReaderState readerState,
             bool isFinalBlock,
             Span<byte> buffer,
             JsonSerializerOptions options,
-            ref ReadStack state)
+            ref ReadStack readStack)
         {
             var reader = new Utf8JsonReader(buffer, isFinalBlock, readerState);
 
+            // If we haven't read in all the data we need to read ahead when reading a json object
+            // or array into a single .NET object so the JsonDocument has all of the needed data
+            // to parse.
+            options.ReadAhead = !isFinalBlock;
```

Comment:
Ideally we only need to do this when we have a custom converter or JsonElement (including the overflow property bag). However, we can optimize that later.

Reply:
The extra seeking is only needed when we don't have the final block and we're reading an object/array into a property value, which I think may not be particularly common? I think we'll have to be somewhat reactionary here when we see real user scenarios that hit this. Not sure how efficiently we can evaluate the serialized object for "simpleness".